### ISSUE TITLE
Data visualization 


### ISSUE THREAD

**Issue by Anjola7077** [Anjola7077/Nova#1]
<!DOCTYPE html>
<html lang="en" ng-app="dataVizApp">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Data Visualization App</title>
    
    
    <script src="https://cdn.tailwindcss.com"></script>

    
    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.8.2/angular.min.js"></script>

    
    <script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/chartjs-plugin-datalabels"></script>
    <script src="https://cdn.jsdelivr.net/npm/chartjs-plugin-zoom"></script>
</head>
<body class="bg-gray-100 p-8" ng-controller="DataVizController">

    <div class="container mx-auto p-6 bg-white rounded shadow-lg">
        
        
        <div class="mb-6">
            <h2 class="text-xl font-bold mb-4">Enter Your Data (comma-separated)</h2>
            <textarea ng-model="inputData" placeholder="Enter data like 10, 20, 30, 40" rows="3" class="w-full p-2 border rounded"></textarea>
        </div>

        
        <div class="mb-6">
            <h2 class="text-lg font-bold mb-4">Enter Labels (comma-separated)</h2>
            <textarea ng-model="inputLabels" placeholder="Enter labels like A, B, C, D" rows="2" class="w-full p-2 border rounded"></textarea>
        </div>

        
        <div class="mb-6">
            <h3 class="text-lg font-semibold mb-2">Select Chart Type</h3>
            <select ng-model="chartType" class="p-2 border rounded w-full">
                <option value="bar">Bar Chart</option>
                <option value="line">Line Chart</option>
                <option value="pie">Pie Chart</option>
                <option value="doughnut">Donut Chart</option>
                <option value="scatter">Scatter Plot</option>
                <option value="area">Area Chart</option>
                <option value="bullet">Bullet Graph</option>
            </select>
        </div>

        
        <div class="mb-6">
            <h3 class="text-lg font-semibold mb-2">Select Chart Color</h3>
            <input type="color" ng-model="chartColor" class="w-full p-2 border rounded">
        </div>

        
        <div class="text-center mb-6">
            <button ng-click="visualizeData()" class="bg-blue-500 text-white px-4 py-2 rounded">Visualize</button>
        </div>

        
        <div class="w-full flex justify-center">
            <div class="w-full md:w-2/3 lg:w-1/2">
                <canvas id="myChart" width="400" height="400"></canvas>
            </div>
        </div>

        
        <div class="text-center mt-6">
            <label for="format-select" class="mr-4">Download as:</label>
            <select id="format-select" ng-model="downloadFormat" class="p-2 border rounded w-1/4">
                <option value="png">PNG</option>
                <option value="jpeg">JPEG</option>
            </select>
            <button ng-click="downloadChart(downloadFormat)" class="bg-green-500 text-white px-4 py-2 rounded ml-4">Download Chart</button>
        </div>

        
        <div class="mt-6 p-4 bg-blue-100 rounded">
            <h4 class="font-semibold mb-2">Instructions:</h4>
            <p>1. Enter your data in the first input as comma-separated numbers (e.g., 10, 20, 30).</p>
            <p>2. Optionally, enter custom labels for your data in the second input (e.g., A, B, C).</p>
            <p>3. Choose a chart type and color, then click "Visualize" to see your chart.</p>
            <p>4. After generating the chart, select a download format and click "Download Chart" to save the image.</p>
        </div>
    </div>

    
    <script>
        angular.module('dataVizApp', [])
        .controller('DataVizController', function($scope) {
            $scope.inputData = '';
            $scope.inputLabels = '';
            $scope.chartType = 'bar';
            $scope.chartColor = '#36A2EB';
            $scope.downloadFormat = 'png';
            let chartInstance;

            // Function to visualize data
            $scope.visualizeData = function() {
                // Validate the input data
                if (!$scope.inputData.match(/^(\d+\s*,\s*)*\d+$/)) {
                    alert('Invalid input. Please enter numbers separated by commas.');
                    return;
                }
                
                let data = $scope.inputData.split(',').map(Number);
                let labels = $scope.inputLabels ? $scope.inputLabels.split(',') : data.map((_, i) => `Label ${i + 1}`);
                let ctx = document.getElementById('myChart').getContext('2d');
                
                // Destroy the previous chart instance if it exists
                if (chartInstance) {
                    chartInstance.destroy();
                }

                let chartConfig = {};

                // Build chart configuration based on selected chart type
                switch ($scope.chartType) {
                    case 'bar':
                        chartConfig = {
                            type: 'bar',
                            data: {
                                labels: labels,
                                datasets: [{
                                    label: 'Data Values',
                                    data: data,
                                    backgroundColor: $scope.chartColor + '80',  // Adjust opacity
                                    borderColor: $scope.chartColor,
                                    borderWidth: 1
                                }]
                            },
                            options: {
                                scales: {
                                    y: {
                                        beginAtZero: true
                                    }
                                }
                            }
                        };
                        break;

                    case 'line':
                        chartConfig = {
                            type: 'line',
                            data: {
                                labels: labels,
                                datasets: [{
                                    label: 'Data Values',
                                    data: data,
                                    borderColor: $scope.chartColor,
                                    backgroundColor: $scope.chartColor + '80',
                                    fill: false,
                                    tension: 0.1
                                }]
                            },
                            options: {
                                scales: {
                                    y: {
                                        beginAtZero: true
                                    }
                                }
                            }
                        };
                        break;

                    case 'doughnut':
                        chartConfig = {
                            type: 'doughnut',
                            data: {
                                labels: labels,
                                datasets: [{
                                    label: 'Data Values',
                                    data: data,
                                    backgroundColor: [
                                        'rgba(255, 99, 132, 0.2)',
                                        'rgba(54, 162, 235, 0.2)',
                                        'rgba(255, 206, 86, 0.2)',
                                        'rgba(75, 192, 192, 0.2)',
                                        'rgba(153, 102, 255, 0.2)'
                                    ],
                                    borderColor: [
                                        'rgba(255, 99, 132, 1)',
                                        'rgba(54, 162, 235, 1)',
                                        'rgba(255, 206, 86, 1)',
                                        'rgba(75, 192, 192, 1)',
                                        'rgba(153, 102, 255, 1)'
                                    ],
                                    borderWidth: 1
                                }]
                            }
                        };
                        break;

                    case 'scatter':
                        chartConfig = {
                            type: 'scatter',
                            data: {
                                datasets: [{
                                    label: 'Scatter Dataset',
                                    data: data.map((val, i) => ({ x: i + 1, y: val })),
                                    backgroundColor: 'rgba(255, 99, 132, 0.2)',
                                    borderColor: 'rgba(255, 99, 132, 1)',
                                    borderWidth: 1
                                }]
                            },
                            options: {
                                scales: {
                                    x: {
                                        type: 'linear',
                                        position: 'bottom'
                                    }
                                }
                            }
                        };
                        break;

                    case 'area':
                        chartConfig = {
                            type: 'line',
                            data: {
                                labels: labels,
                                datasets: [{
                                    label: 'Area Chart',
                                    data: data,
                                    borderColor: $scope.chartColor,
                                    backgroundColor: $scope.chartColor + '80',
                                    fill: true,  // Fill the area below the line 
                    tension: 0.4
                                }]
                            },
                            options: {
                                scales: {
                                    y: {
                                        beginAtZero: true
                                    }
                                }
                            }
                        };
                        break;

                    case 'bullet':
                        chartConfig = {
                            type: 'bar',
                            data: {
                                labels: ['Target'],
                                datasets: [
                                    {
                                        label: 'Target',
                                        data: [100],  // Example target value
                                        backgroundColor: 'rgba(200, 200, 200, 0.6)',  // Target bar color
                                        borderWidth: 1
                                    },
                                    {
                                        label: 'Progress',
                                        data: [data[0]],  // The first value is considered progress
                                        backgroundColor: $scope.chartColor,
                                        borderWidth: 1
                                    }
                                ]
                            },
                            options: {
                                scales: {
                                    x: {
                                        stacked: true
                                    },
                                    y: {
                                        beginAtZero: true,
                                        max: 100  // Example max value for the bullet chart
                                    }
                                }
                            }
                        };
                        break;

                    default:
                        alert('Unsupported chart type');
                        return;
                }

                // Create the new chart
                chartInstance = new Chart(ctx, chartConfig);
            };

            // Function to download the chart as an image
            $scope.downloadChart = function(format = 'png') {
                let canvas = document.getElementById('myChart');
                let link = document.createElement('a');
                link.href = canvas.toDataURL(`image/${format}`);
                link.download = `chart.${format}`;
                link.click();
            };
        });
    </script>

</body>
</html>
